### PR TITLE
Week 19 - Project Happy Thoughts API, Isabella Markskog

### PR DESCRIPTION
**Pull request by IsabellaM5**
**Link to my deployed project**
- Link to my deployed API (this repo): https://project-happy-thoughts-isabm5.herokuapp.com/ 
- Link to my deployed fullstack application: https://heuristic-bassi-e1a3f1.netlify.app/ 
- Link to my repo for the frontend: https://github.com/IsabellaM5/project-happy-thoughts 

**My reflections on how this week's project turned out**
I did all the blue and red level requirements. I also implemented the feature to be able to edit an already existing thought by making a PATCH request, which I'd say is black level.

I think this week's project was great! Super fun to finally connect the dots and see my very first fullstack application come to life 😍 There were a lot of things I would've liked to add, for examples see my README file, but I was a bit short on time this week.

**Things I'd like to have clarified or explained in more detail**
Nothing at the moment. I've been enjoying thoroughly the backend parts so far 😍

### CODE REVIEW

**Review by ElleFrankenberg — APPROVED** [review]
Wow, it really shows that you thought that this week was great! You have definitely for filled all of the blue requirements and so much more. I love the heart loader and the new background image and all of the new functionalities you have added to your site. Everything is working, clean code and nice documentation, Great job!!💥

File: server.js [hunk 1]
```diff
@@ -1,30 +1,164 @@
...
+
+const mongoUrl = process.env.MONGO_URL || "mongodb://localhost/happyThoughts";
+mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true });
+mongoose.Promise = Promise;
+
+const thoughtSchema = new mongoose.Schema({
+  message: {
+    type: String,
+    required: true,
+    unique: true,
```
Comment (KarolCislowski) [inline server.js:14]: The message is not a good field to make it unique. if 3 recruiters come and send "testing" 2 of them will think it is not working 😭

File: server.js [hunk 2]
```diff
@@ -1,30 +1,164 @@
...
+app.use(cors());
+app.use(express.json());
+
+app.use((req, res, next) => {
+  if (mongoose.connection.readyState === 1) {
+    next();
+  } else {
+    res.status(503).json({ err: 'Service unavailable' });
+  }
+});
```
Comment (KarolCislowski) [inline server.js:50]: 🥇

File: server.js [hunk 3]
```diff
@@ -1,30 +1,164 @@
...
+    }
+    res.status(400).json({ message: 'Could not save the thought to the database.', err });
+  }
+});
+
+app.post('/thoughts/:thoughtId/like', async (req, res) => {
+  const { thoughtId } = req.params;
+
+  try {
+    const updatedHeart = await Thought.findByIdAndUpdate(thoughtId, { $inc: { hearts: 1 } }, { new: true });
```
Comment (KarolCislowski) [inline server.js:105]: 💯

File: server.js [hunk 4]
```diff
@@ -1,30 +1,164 @@
...
+  try {
+    res.send(listEndpoints(app));
+  } catch (err) {
+    res.status(404).json({ err: 'Page not found' });
+  }
+});
+
+app.get('/thoughts', async (req, res) => {
+  try {
+    const thoughts = await Thought.find().sort({ createdAt: 'desc' }).limit(20).exec();
```
Comment (KarolCislowski) [inline server.js:63]: 👍🏻

File: server.js [hunk 5]
```diff
@@ -1,30 +1,164 @@
...
+    const updatedMessage = await Thought.findOneAndReplace({ _id: thoughtId }, req.body, { new: true });
+    if (updatedMessage) {
+      res.json(updatedMessage);
+    } else {
+      res.status(404).json({ message: 'Could not find the thought in the database.' });
+    }
+  } catch (err) {
+    res.status(400).json({ message: 'Could not update the thought in the database.', err });
+  }
+});
```
Comment (KarolCislowski) [inline server.js:160]: Remember to restrict access to these endpoints because not all users should have access to these functionalities 🤓

**Review by KarolCislowski — APPROVED** [review]
Great API but rethink usage of unique in this case 🤓

**Pull request closed without merging**